### PR TITLE
vscode-extensions.ms-dotnettools.csharp: 1.23.16 -> 1.25.0

### PR DESCRIPTION
**Pull request by GGG-KILLER**
###### Description of changes

All combined changes from the following:
- [v1.23.17](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.17)
- [v1.24.0](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.0)
- [v1.24.1](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.1)
- [v1.24.2](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.2)
- [v1.24.3](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.3)
- [v1.24.4](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.4)
- [v1.25.0](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.25.0)

Notable changes:
- Extension no longer ships with Mono nor MSBuild so you'll need the .NET SDK installed on your system (at least .NET 6 as that's what OmniSharp uses)
- Extension vsix is now fetched from GitHub as vsassets.io was returning an outdated version for v1.25.0
- Binary dependencies are no longer fetched as they come bundled with the extension
- Support for ARM64 on Linux and Darwin was added

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 1]
```diff
@@ -1,65 +1,74 @@
...
-    (rtDepBinSrcByName "Debugger")
-  ] ++ lib.optionals (stdenv.isDarwin) [
-  # Include the aarch64-darwin debugger binaries on x86_64-darwin.  Even though OmniSharp will be
-  # running under Rosetta 2, debugging will fail to start if both sets of binaries are not present.
-    (rtDepsBinSrcs."Debugger__aarch64-darwin")
-  ];
-  razor = rtDepBinSrcByName "Razor";
+  vsixInfo = {
+    x86_64-linux = {
+      url = "https://github.com/OmniSharp/omnisharp-vscode/releases/download/v1.25.0/csharp-1.25.0-linux-x64.vsix";
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:12]: Please substitute the version numbers.

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 2]
```diff
@@ -1,65 +1,74 @@
...
+        ".razor/rzls"
+      ];
+    };
+    aarch64-linux = {
+      url = "https://github.com/OmniSharp/omnisharp-vscode/releases/download/v1.25.0/csharp-1.25.0-linux-arm64.vsix";
+      sha256 = "0nsjgrb7y4w71w1gnrf50ifwbmjidi4vrw2fyfmch7lgjl8ilnhd";
+      binaries = [
+        ".debugger/vsdbg-ui"
+        ".debugger/vsdbg"
+        ".omnisharp/1.39.0-net6.0/OmniSharp"
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:28]: this should also be in one variable

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 3]
```diff
@@ -1,65 +1,74 @@
...
+        ".debugger/arm64/vsdbg-ui"
+        ".debugger/arm64/vsdbg"
+        ".debugger/x86_64/vsdbg-ui"
+        ".debugger/x86_64/vsdbg"
+        ".omnisharp/1.39.0-net6.0/OmniSharp"
+        ".razor/createdump"
+        ".razor/rzls"
+      ];
+    };
+  }.${stdenv.targetPlatform.system};
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:56]: missing or throw unsupported platform

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 4]
```diff
@@ -78,23 +87,11 @@ vscode-utils.buildVscodeMarketplaceExtension {
...
-      else
-        1>&2 echo "ERROR: unzip('$?')."
-      fi
-    }
-
     patchelf_add_icu_as_needed() {
       declare elf="''${1?}"
       declare icu_major_v="${
-        with builtins; head (splitVersion (parseDrvName icu.name).version)}"
+      with builtins; head (splitVersion (parseDrvName icu.name).version)
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:93]:
```suggestion
      lib.head (lib.splitVersion (lib.parseDrvName icu.name).version)
```

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 5]
```diff
@@ -107,46 +104,26 @@ vscode-utils.buildVscodeMarketplaceExtension {
       patchelf_add_icu_as_needed "$elf"
       patchelf --add-needed "libssl.so" "$elf"
       patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
-        --set-rpath "${lib.makeLibraryPath [ stdenv.cc.cc openssl icu.out ]}:\$ORIGIN" \
+        --set-rpath "${lib.makeLibraryPath [stdenv.cc.cc openssl icu.out]}:\$ORIGIN" \
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:107]:
```suggestion
        --set-rpath "${lib.makeLibraryPath [ stdenv.cc.cc openssl icu.out ]}:\$ORIGIN" \
```

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 6]
```diff
@@ -107,46 +104,26 @@ vscode-utils.buildVscodeMarketplaceExtension {
...
-  '' + lib.optionalString stdenv.isLinux ''
-    patchelf_common "$vsdbg_dir/vsdbg"
-    patchelf_common "$vsdbg_dir/vsdbg-ui"
-    patchelf_common "$razor_dir/rzls"
-
-  '' + lib.optionalString stdenv.isDarwin ''
-    substituteInPlace $omnisharp_dir/etc/config \
-      --replace "libmono-native-compat.dylib" "libmono-native.dylib"
-  '';
+  '' + (builtins.concatStringsSep "\n" (map
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:111]:
```suggestion
  '' + (lib.concatStringsSep "\n" (map
```

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix [hunk 7]
```diff
@@ -107,46 +104,26 @@ vscode-utils.buildVscodeMarketplaceExtension {
...
+    (bin: ''
+      chmod +x "${bin}"
+    '')
+    vsixInfo.binaries))
+  + lib.optionalString stdenv.isLinux (builtins.concatStringsSep "\n" (map
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:116]:
```suggestion
  + lib.optionalString stdenv.isLinux (lib.concatStringsSep "\n" (map
```

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix (hunk 3 above)
Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:56]: I'm unsure on how to do this. Would I do something like `or error "Error message"`?

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix (hunk 2 above)
Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:28]: This version is the one bundled with the extension, not the one in nixpkgs

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix (hunk 5 above)
Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:107]: yeah sorry, my formatOnSave undid that spacing

File: pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:32]: yeah I know but we repeat that for every download. Would be nicer if we didn't.

Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:32]: I mean, if we go that route, one could argue that every binary in this list could be in a variable then we combine them in the arrays

Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix:32]: However what I think I'll do is create sets of binaries and combine them based on which ones are available

Comment (GGG-KILLER) [reply pkgs/applications/editors/vscode/extensions/ms-dotnettools-csharp/default.nix]: @SuperSandro2000 I've updated the PR with the approach I mentioned above, could you check it?

**Pull request merged**